vm_test.go

```diff
@@ -594,3 +595,81 @@ func TestStashMemUsage(t *testing.T) {
 		})
 	}
 }
+
+func TestTickTracking(t *testing.T) {
+	tests := []struct {
+		name                        string
+		script                      string
+		functionTickTrackingEnabled bool
+		expectedTickMetrics         map[string]uint64
+	}{
+		{
+			name: "should track looping function ticks when tick tracking is enabled",
+			script: `
+				function f() {
+					for (var i = 0; i < 100; i++) {
+					}
+				}
+				f()
+			`,
+			functionTickTrackingEnabled: true,
+			expectedTickMetrics:         map[string]uint64{"test.js_": 6, "test.js_f": 809},
+		},
+		{
+			name: "should track larger looping function ticks when tick tracking is enabled",
+			script: `
+				function f() {
+					for (var i = 0; i < 1000; i++) {
+					}
+				}
+				f()
+			`,
+			functionTickTrackingEnabled: true,
+			expectedTickMetrics:         map[string]uint64{"test.js_": 6, "test.js_f": 8009},
+		},
+		{
+			name: "should track fib function ticks when tick tracking is enabled",
+			script: `
+				function fib(n) {
+					if (n < 2) return n;
+					return fib(n - 2) + fib(n - 1);
+				}
+				fib(35);
+			`,
+			functionTickTrackingEnabled: true,
+			expectedTickMetrics:         map[string]uint64{"test.js_": 7, "test.js_fib": 358328431},
+		},
+		{
+			name: "should not track function ticks when tick tracking is disabled",
+			script: `
+				function f() {
+					for (var i = 0; i < 100; i++) {
+					}
+				}
+				f()
+			`,
+			functionTickTrackingEnabled: false,
+			expectedTickMetrics:         map[string]uint64{},
+		},
+	}
+
+	for _, tc := range tests {
```

Comment:
> [q1]can we add some more tests such as executing executing class functions and multiple nested functions?

Yeah absolutely!

> [q2] Also, do you mind sharing the metrics that we would generate from baas when using a dependency such as the aws sdk? The easiest way to do this might be to run our dependencies tests in evergreen with the tracking enabled.

Would that data be in splunk? Or would we need to parse the output of the tests?

Reply:
[Evergreen task](https://spruce.mongodb.com/version/65bd532a1e2d17419332be0c/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) with tick metric tracking hardcoded to true. Will update this thread with the results.

Reply:
Hmm for some reason the logs aren't showing the "tick metrics" logs for that evergreen run.

I think my local testing might alleviate some of your concerns though.

```
exports = async function(arg){
  const { Buffer } = require('buffer');
  const loremIpsumText = // long string
  return Buffer.from(loremIpsumText);
};
```

![image](https://github.com/mongodb-forks/goja/assets/6693710/6d7f358b-e7ee-48ac-8d39-876e92f81603)

The majority of usage for this Buffer.from example is BufferJS::utf8ToBytes and BufferJS::blitBuffer.

Reply:
> Hmm for some reason the logs aren't showing the "tick metrics" logs for that evergreen run.

I have a feeling this is happening because of how tick metrics and aggregated and published asynchronously. It doesn't even look like the cmd/server/main.go executes which is where this asynchronous logic would start.

I went ahead and set this up manually though, here are the results when running against [aws-sdk-v3/s3](https://github.com/10gen/baas/blob/165ae68dc53d625d415347bf1703616f7f291f80/function/execution/vm/dependencies_async/dependencies_async_test.go#L299)

![image](https://github.com/mongodb-forks/goja/assets/6693710/29450f3c-6bc8-4f60-93af-0c2484e868b2)

Reply:
hmac lol

Reply:
oh nice work, hello mr Buffer not surprised to see you up there!